url1d5e9d990f_Telsis_jOCP_tgz-pJ8-com_telsis_jocp_sampleApp_BasicOCPAppJ8/src/com/telsis/jocp/OCPSystemManager.java

```diff
@@ -0,0 +1,1344 @@
+        }
+
+        /**
+         * Raise a link alarm on the current link.
+         */
+        public synchronized void raiseLinkAlarm() {
+            if (!cleanedUp) {
+                alarms.raiseAlarm(ALARM_LINK_NOT_AVAILABLE, this, getLinkID(),
+                        new String[] {toString()});
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "Link(" + linkManager + ")";
+        }
+
+        @Override
+        public void congestionEvent(
+                final EnumMap<CongestionType, Integer> congestionList) {
+            // Not used, but required by ServerEventNotifier
+        }
+
+        @Override
+        public void reloadConfiguration() {
+            // Not used, but required by ServerEventNotifier
+        }
+
+        @Override
+        public void stateEvent(final ActiveStates activeState) {
+            // Not used, but required by ServerEventNotifier
+        }
+    }
+
+    /**
+     * Management thread for the OCP system manager. This thread periodically
+     * manages the following:
+     * <ul>
+     * <li>Promotion of slave links.</li>
+     * <li>Gapping updates.</li>
+     * <li>Suspect link detection.</li>
+     * <li>Preferred unit selection.</li>
+     * <li>System state.</li>
+     * </ul>
+     *
+     * @author Telsis
+     */
+    private class SystemThread implements Runnable {
+        /** The thread that this object is running in. */
+        private Thread  myThread;
+        /** A flag to detect when this thread should be shut down. */
+        private boolean shutdownThread;
+
+        /**
+         * Create a new stopped SystemThread.
+         */
+        public SystemThread() {
+            super();
+
+            shutdownThread = false;
+            myThread = null;
+        }
+
+        /**
+         * Entry point for the thread. This function should not be called
+         * directly. On start-up, the thread brings up the OCP links and then
+         * loops updating the system status at regular intervals. When a
+         * shut down request is detected, the OCP links are disconnected and the
+         * thread terminates.
+         */
+        @Override
+        public void run() {
+            myThread = Thread.currentThread();
+            log.debug("SystemThread: thread starting");
+            watchdog.start();
+
+            synchronized (this) {
+                state = recalcMode();
+
+                // Bring up the OCP Links
+                for (Link link : links) {
+                    link.linkManager.connect(watchdog);
+                }
+
+                while (true) {
+                    watchdog.pat();
+                    internalRun();
+
+                    if (shutdownThread) {
+                        break;
+                    }
+
+                    try {
+                        this.wait(1 * MILLISECONDS);
+                    } catch (InterruptedException e) { // CSIGNORE: EmptyBlock
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)